File: hermit-cli/tests/prodcons_determinism.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! End-to-end determinism coverage for a blocking producer/consumer guest.
//!
//! `tests/c/prodcons_determinism.c` runs several producer and consumer threads
//! against a fixed-capacity ring guarded by a mutex and two condition variables
//! (not-full / not-empty). Unlike a spin-contention probe, the threads block on
//! the condition variables, so Hermit must serialize the futex-backed
//! mutex/condvar wakeups deterministically. The guest folds every consumed
//! payload into an order-independent checksum and verifies the produced and
//! consumed counts, then prints the counts and the `prodcons success` marker,
//! so a stable stdout is the determinism witness.

use std::fs;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;
use std::process::Output;
use std::sync::Mutex;
use std::sync::MutexGuard;
use std::sync::OnceLock;

const RUNS: usize = 5;
const SUCCESS_MARKER: &str = "prodcons success\n";

static HERMIT_RUN_LOCK: Mutex<()> = Mutex::new(());
static PRODCONS_GUEST: OnceLock<PathBuf> = OnceLock::new();

fn hermit_run_lock() -> MutexGuard<'static, ()> {
    HERMIT_RUN_LOCK
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner())
}

fn command_output(mut command: Command, label: &str) -> Output {
    let rendered = format!("{command:?}");
    let output = command
        .output()
        .unwrap_or_else(|error| panic!("failed to start {label}: {rendered}: {error}"));
    assert!(
        output.status.success(),
        "{label} failed: {rendered}\nstatus: {}\nstdout:\n{}\nstderr:\n{}",
        output.status,
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr),
    );
    output
}

fn prodcons_guest() -> &'static Path {
    PRODCONS_GUEST.get_or_init(|| {
        let repository = Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .expect("hermit-cli should be inside the repository");
        let build_root = Path::new(env!("CARGO_TARGET_TMPDIR")).join("prodcons-determinism");
        fs::create_dir_all(&build_root).expect("failed to create prodcons guest build directory");
        let output = build_root.join("prodcons_determinism");

        let mut command = Command::new("cc");
        command
            .args([
                "-O0",
                "-g",
                "-D_GNU_SOURCE",
                "-std=c11",
                "-Wall",
                "-Wextra",
                "-Werror",
                "-pthread",
            ])
            .arg(repository.join("tests/c/prodcons_determinism.c"))
            .arg("-o")
            .arg(&output);
        command_output(command, "prodcons guest compilation");
        output
    })
}

fn run_once(run: usize) -> Vec<u8> {
    let mut command = Command::new(env!("CARGO_BIN_EXE_hermit"));
    command
        .args([
            "run",
            "--base-env=minimal",
            "--no-virtualize-cpuid",
            "--max-timeslice=disabled",
            "--",
        ])
        .arg(prodcons_guest());

    let output = command_output(command, &format!("prodcons run {run}/{RUNS}"));
    assert!(
        output.stdout.ends_with(SUCCESS_MARKER.as_bytes()),
        "prodcons omitted its success marker on run {run}/{RUNS}:\n{}",
        String::from_utf8_lossy(&output.stdout),
    );
    output.stdout
}

#[test]
fn prodcons_counts_and_checksum_are_deterministic() {
    let _guard = hermit_run_lock();
    let expected = run_once(1);

    for run in 2..=RUNS {
        let actual = run_once(run);
        assert_eq!(
            actual,
            expected,
            "prodcons output changed on run {run}/{RUNS}:\nexpected:\n{}actual:\n{}",
            String::from_utf8_lossy(&expected),
            String::from_utf8_lossy(&actual),
        );
    }
}

#[test]
#[ignore = "e2e: requires hermit + mount namespaces"]
fn prodcons_reaches_strict_verify_l2() {
    let _guard = hermit_run_lock();
    let mut command = Command::new("timeout");
    command
        .args(["--kill-after", "10s", "120s"])
        .arg(env!("CARGO_BIN_EXE_hermit"))
        .args([
            "--log=info",
            "run",
            "--strict",
            "--verify",
            "--no-virtualize-cpuid",
            "--preemption-timeout=disabled",
            "--",
        ])
        .arg(prodcons_guest());

    let output = command_output(command, "prodcons strict verification");
    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    assert!(
        stdout.contains("Determinism verified") || stderr.contains("Determinism verified"),
        "prodcons exited 0 without Hermit's determinism marker\nstdout:\n{stdout}\nstderr:\n{stderr}"
    );
}

File: tests/c/prodcons_determinism.c
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

// Bounded-queue producer/consumer determinism guest. Multiple producers and
// consumers share a fixed-capacity ring guarded by a mutex and two condition
// variables (not-full / not-empty). A fixed amount of work flows through the
// queue and each consumed payload is folded into an order-independent checksum.
// Hermit must serialize the mutex/condvar wakeups deterministically, so the
// printed counts and checksum are identical across runs. This is distinct from
// a spin-contention probe: the threads block on condition variables rather than
// busy-waiting. Prints "prodcons success" as the final line on success.
#include <pthread.h>
#include <stdint.h>
#include <stdio.h>

#define CAP 8
#define NPROD 4
#define NCONS 3
#define PER_PROD 500
#define TOTAL (NPROD * PER_PROD)

static int buf[CAP];
static int head, tail, count;
static pthread_mutex_t mu = PTHREAD_MUTEX_INITIALIZER;
static pthread_cond_t not_full = PTHREAD_COND_INITIALIZER;
static pthread_cond_t not_empty = PTHREAD_COND_INITIALIZER;

static int produced_total, consumed_total;
static uint64_t consumed_sum;

static void *producer(void *arg) {
    long id = (long)arg;
    for (int i = 0; i < PER_PROD; i++) {
        int item = (int)(id * 100000 + i);
        pthread_mutex_lock(&mu);
        while (count == CAP) {
            pthread_cond_wait(&not_full, &mu);
        }
        buf[tail] = item;
        tail = (tail + 1) % CAP;
        count++;
        produced_total++;
        pthread_cond_signal(&not_empty);
        pthread_mutex_unlock(&mu);
    }
    return NULL;
}

static void *consumer(void *arg) {
    (void)arg;
    for (;;) {
        pthread_mutex_lock(&mu);
        while (count == 0 && consumed_total < TOTAL) {
            pthread_cond_wait(&not_empty, &mu);
        }
        if (count == 0 && consumed_total >= TOTAL) {
            pthread_mutex_unlock(&mu);
            break;
        }
        int item = buf[head];
        head = (head + 1) % CAP;
        count--;
        consumed_total++;
        consumed_sum += (uint64_t)(uint32_t)item;
        pthread_cond_signal(&not_full);
        pthread_cond_broadcast(&not_empty);
        pthread_mutex_unlock(&mu);
    }
    return NULL;
}

int main(void) {
    pthread_t p[NPROD], c[NCONS];
    for (long i = 0; i < NPROD; i++) {
        pthread_create(&p[i], NULL, producer, (void *)i);
    }
    for (long i = 0; i < NCONS; i++) {
        pthread_create(&c[i], NULL, consumer, (void *)i);
    }
    for (int i = 0; i < NPROD; i++) {
        pthread_join(p[i], NULL);
    }
    pthread_mutex_lock(&mu);
    pthread_cond_broadcast(&not_empty);
    pthread_mutex_unlock(&mu);
    for (int i = 0; i < NCONS; i++) {
        pthread_join(c[i], NULL);
    }

    uint64_t expect = 0;
    for (long id = 0; id < NPROD; id++) {
        for (int i = 0; i < PER_PROD; i++) {
            expect += (uint64_t)(uint32_t)(int)(id * 100000 + i);
        }
    }

    printf("produced=%d\n", produced_total);
    printf("consumed=%d\n", consumed_total);
    printf("consumed_sum=%llu\n", (unsigned long long)consumed_sum);
    if (produced_total != TOTAL || consumed_total != TOTAL || consumed_sum != expect) {
        printf("prodcons MISMATCH expect_sum=%llu\n", (unsigned long long)expect);
        return 1;
    }
    printf("prodcons success\n");
    return 0;
}
